### PR TITLE
[1.12] Author backport of #29239 "Dns proxy use original source address and port"

### PR DESCRIPTION
**Pull request by jrajahalme**
* [ ] #29239 (@jrajahalme)

Once this PR is merged, a GitHub action will update the labels of these PRs:
```upstream-prs
29239
```

### CODE REVIEW

**Comment by jrajahalme** [comment]
/test-1.19-4.9

**Comment by jrajahalme** [comment]
Rebased to pick up Ci fixes.

**Comment by jrajahalme** [comment]
/test-backport-1.12

File: pkg/fqdn/dnsproxy/proxy.go [hunk 1]
```diff
@@ -594,27 +598,106 @@ func (p *DNSProxy) CheckAllowed(endpointID uint64, destPort uint16, destID ident
...
 		return fmt.Errorf("error setting SO_MARK: %w", err)
 	}
+
+	// Rest of the options are only set in the transparent mode.
+	if !option.Config.DNSProxyEnableTransparentMode {
+		return nil
+	}
+
+	// Set IP_TRANSPARENT to be able to use a non-host address as the source address
+	if err := transparentSetsockopt(fd, ipv4, !ipv4); err != nil {
```
Comment (pippolo84) [inline pkg/fqdn/dnsproxy/proxy.go:657]: Same [comment](https://github.com/cilium/cilium/pull/30216/files#r1452525997) as for v1.13 backport: Differently from v1.14 onwards, here we are calling `transparentSetsockopt` and thus also setting `IP_RECVORIGDSTADDR`. I don't think this makes any actual difference, but maybe it is better to just set `IP_TRANSPARENT` or `IPV6_TRANSPARENT` according to the `ipv4` flag.

**Review by dylandreimerink — APPROVED** [review]

**Comment by jrajahalme** [comment]
/test-1.16-4.9

**Pull request merged**